heartbeat/autodiscover/builder/hints/monitors.go

```diff
@@ -0,0 +1,215 @@
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package hints
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/beats/libbeat/autodiscover"
+	"github.com/elastic/beats/libbeat/autodiscover/builder"
+	"github.com/elastic/beats/libbeat/autodiscover/template"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+func init() {
+	autodiscover.Registry.AddBuilder("hints", NewHeartbeatHints)
+}
+
+const (
+	montype    = "type"
+	schedule   = "schedule"
+	hosts      = "hosts"
+	processors = "processors"
+)
+
+type heartbeatHints struct {
+	config *config
+	logger *logp.Logger
+}
+
+// NewHeartbeatHints builds a heartbeat hints builder
+func NewHeartbeatHints(cfg *common.Config) (autodiscover.Builder, error) {
+	config := defaultConfig()
+	err := cfg.Unpack(config)
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to unpack hints config due to error: %v", err)
+	}
+
+	return &heartbeatHints{config, logp.NewLogger("hints.builder")}, nil
+}
+
+// Create config based on input hints in the bus event
+func (hb *heartbeatHints) CreateConfig(event bus.Event) []*common.Config {
+	var hints common.MapStr
+	hIface, ok := event["hints"]
+	if ok {
+		hints, _ = hIface.(common.MapStr)
+	}
+
+	monitorConfig := hb.getRawConfigs(hints)
+
+	// If explicty disabled, return nothing
+	if builder.IsDisabled(hints, hb.config.Key) {
+		hb.logger.Warnf("heartbeat config disabled by hint: %+v", event)
+		return []*common.Config{}
+	}
+
+	port, _ := common.TryToInt(event["port"])
+
+	host, _ := event["host"].(string)
+	if host == "" {
+		return []*common.Config{}
+	}
+
+	if monitorConfig != nil {
+		configs := []*common.Config{}
+		for _, cfg := range monitorConfig {
+			if config, err := common.NewConfigFrom(cfg); err == nil {
+				configs = append(configs, config)
+			}
+		}
+		hb.logger.Debugf("generated config %+v", configs)
+		// Apply information in event to the template to generate the final config
+		return template.ApplyConfigTemplate(event, configs)
+	}
+
+	tempCfg := common.MapStr{}
+	monitors := hb.getMonitors(hints)
+
+	var configs []*common.Config
+	for _, monitor := range monitors {
+		// Every monitor must have a type associated with it
+		if _, ok := monitor[montype]; !ok {
+			hb.logger.Warn("type of monitor is mandatory")
```

Comment:
I don't think we actually need this check here, this should be caught at a higher level where we can also add more context. https://github.com/elastic/beats/pull/13675/files should improve things.